### PR TITLE
Fix endless loading thread report header

### PR DESCRIPTION
**Pull request by tsa321**
### Fixed Issues

$ https://github.com/Expensify/App/issues/40928
PROPOSAL: https://github.com/Expensify/App/issues/40928#issuecomment-2095619165

### Tests

1. Open admins or announce room
2. Send a message containing only mentioned room name (for example #admins or #announce)
3. Select `Reply in thread` of that message
4. Wait for at least 5 seconds
5. Verify the thread report header looks like common thread header and is not changed into endless loading
- [x] Verify that no errors appear in the JS console

### QA Steps

1. Open admins or announce room
2. Send a message containing only mentioned room name (for example #admins or #announce)
3. Select `Reply in thread` of that message
4. Wait for at least 5 seconds
5. Verify the thread report header looks like common thread header and is not changed into endless loading
- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/114975543/88af58a2-b726-455d-813b-2eaef8ff3131

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/114975543/ac497bb4-a2d8-49f0-bf88-0b5a02f49c94

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/114975543/06e19ed4-d711-497b-8f23-b9571aaf59c6

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/114975543/115871e5-03d1-45c0-a5b3-e4b614d07345

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/114975543/ebecd9d2-dc13-41bc-9f6a-262f3970accd

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/114975543/3aac6287-f148-4b61-a4ec-941d2ecc6ede

</details>

### CODE REVIEW

File: src/libs/ReportUtils.ts
Comment (tsa321) [inline src/libs/ReportUtils.ts:3096]: I am using ExpensiMark.htmltoText instead of my original code.
But there is typecheck error for `accountIdToName` of extras in [here](https://github.com/Expensify/expensify-common/blob/84ac0394ab1021d3d892338c4dbb3f87c05b0bc9/lib/ExpensiMark.d.ts#L39). The field must be `accountIdToName` and not `accountIDToName`. 
The typecheck will fails if I include `accountIdToName` in this line.

File: src/libs/ReportUtils.ts
Comment (tsa321) [inline src/libs/ReportUtils.ts:3087]: Instead of searching for mentioned room name id before  parsing, we could pass callback `(id) => getReportName(getReport(id))` into extras. In [here](https://github.com/Expensify/expensify-common/blob/84ac0394ab1021d3d892338c4dbb3f87c05b0bc9/lib/ExpensiMark.js#L536). 

But this will need changes in ExpensiMark.js

**Comment by tsa321** [comment]
@mollfpr PR is ready for review....

**Comment by tsa321** [comment]
Also, since thread report will be displayed in LHN, the function `parseHtmlToString` will be used many times when rendering LHN. Even when sending a message, the LHN will be rendered. To optimize the performance we could cache the result into object with key `reportActionID_lastModified`. So, as long as report action isn't modified we will just return the value of that key.

File: src/libs/ReportUtils.ts
Comment (luacmartins) [inline src/libs/ReportUtils.ts:3082]:
```suggestion
    const reportIDToName: Record<string, string> = {};
```

File: src/libs/ReportUtils.ts
Comment (luacmartins) [inline src/libs/ReportUtils.ts:3091]:
```suggestion
    const accountIDToName: Record<string, string> = {};
```

File: src/libs/ReportUtils.ts
Comment (luacmartins) [inline src/libs/ReportUtils.ts:3092]:
```suggestion
    const accountIDs = Array.from(html.matchAll(mentionUserRegex), (mention) => Number(mention[1]));
```

File: src/libs/ReportUtils.ts
Comment (luacmartins) [inline src/libs/ReportUtils.ts:490]: Let's leave a comment explaining why we need this cache map

File: src/libs/ReportUtils.ts
Comment (luacmartins) [reply src/libs/ReportUtils.ts:3097]: We need to include `accountIdToName`, don't we?

**Review by luacmartins — CHANGES_REQUESTED** [review]
Looking good overall. Left a few comments though.

**Comment by luacmartins** [comment]
Bump for review @mollfpr

**Comment by mollfpr** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

File: src/libs/ReportUtils.ts
Comment (tsa321) [reply src/libs/ReportUtils.ts:3082]: @luacmartins I am following the ExpensiMark field name in [here](https://github.com/Expensify/expensify-common/blob/c3cc2c0d07c961df08a7147b9c4405014521b8bd/lib/ExpensiMark.js#L539).

If I use reportIDToName it won't work.

File: src/libs/ReportUtils.ts
Comment (tsa321) [reply src/libs/ReportUtils.ts:3091]: and also with `accountIdToName` in [here](https://github.com/Expensify/expensify-common/blob/c3cc2c0d07c961df08a7147b9c4405014521b8bd/lib/ExpensiMark.js#L552C53-L552C68)

File: src/libs/ReportUtils.ts
Comment (tsa321) [reply src/libs/ReportUtils.ts:3097]: @luacmartins  Yes, but there will be lint error if I used `accountIdToName`. 
the field is defined as `accountIDName` in [here](https://github.com/Expensify/expensify-common/blob/84ac0394ab1021d3d892338c4dbb3f87c05b0bc9/lib/ExpensiMark.d.ts#L39).

But in the the javascript function it uses `accountIdName` with lower `d` in [here](https://github.com/Expensify/expensify-common/blob/c3cc2c0d07c961df08a7147b9c4405014521b8bd/lib/ExpensiMark.js#L552C53-L552C68).

Should I make changes in `ExpensiMark` ?

Comment (luacmartins) [reply src/libs/ReportUtils.ts:3097]: Yes, let's do it.

**Comment by tsa321** [comment]
@mollfpr I have made a PR in expensify common: https://github.com/Expensify/expensify-common/pull/708

**Review by luacmartins — COMMENTED** [review]
Waiting on the expensify-common PR to be merged so we can bump up the version here and test this again

**Comment by luacmartins** [comment]
@tsa321 I merged the `expensify-common` PR. Can you bump up the version on this one so we can continue with testing please?

**Comment by tsa321** [comment]
@luacmartins @mollfpr the new release of  Expensify-Common will be published to npm (I think by this https://github.com/Expensify/expensify-common/pull/699 ). After that commit, the expensify-common lib folder won't be included in expensify-common folder inside node_modules. `import ExpensiMark from 'expensify-common/lib/ExpensiMark';` will be modified into `import ExpensiMark from 'expensify-common/ExpensiMark`

and PR: https://github.com/Expensify/App/pull/42387 will migrate all import usage of expensify-common.
I think we should wait for https://github.com/Expensify/App/pull/42387

**Comment by blazejkustra** [comment]
>  After that commit, the expensify-common lib folder won't be included in expensify-common folder inside node_modules. `import ExpensiMark from 'expensify-common/lib/ExpensiMark';` will be modified into `import ExpensiMark from 'expensify-common/ExpensiMark`

[PR](https://github.com/Expensify/App/pull/42387) that upgrades expensify-common is ready. Just to clarify it won't be `import ExpensiMark from 'expensify-common/ExpensiMark` but instead:
```ts
import {ExpensiMark} from 'expensify-common';
//or
import ExpensiMark from 'expensify-common/dist/ExpensiMark';
```

**Comment by luacmartins** [comment]
Bump for review @mollfpr

**Comment by luacmartins** [comment]
@tsa321 can you update the PR to use the new expensify-common version?

**Comment by tsa321** [comment]
@luacmartins I think we are holding on https://github.com/Expensify/App/pull/42387
After it is merged, I will merge main.

**Review by luacmartins — COMMENTED** [review]
Still holding on https://github.com/Expensify/App/pull/42387

**Comment by blazejkustra** [comment]
@tsa321 @luacmartins https://github.com/Expensify/App/pull/42387 was just merged 🚀

**Comment by luacmartins** [comment]
Nice! Let's continue work on this PR @tsa321

**Comment by tsa321** [comment]
@mollfpr I have merged main. PR is ready for review...

**Review by mollfpr — APPROVED** [review]
LGTM 👍 

All yours @luacmartins

**Review by luacmartins — APPROVED** [review]
LGTM

**Pull request merged**
